### PR TITLE
Add compatibility tests for ConvertTo-Json

### PR DESCRIPTION
**Pull request by yotsuda**
# PR Summary

Add compatibility tests for `ConvertTo-Json` to improve test coverage.

## PR Context

This PR adds tests to validate `ConvertTo-Json` behavior, extracted from #26637 as suggested by @iSazonov.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] These changes were tested
- [x] Test added (test-only PR)

## Tests Added

The following tests are added to `ConvertTo-Json.Tests.ps1`:

| Test | Description |
|------|-------------|
| Uri serialization | Validates Uri objects serialize correctly |
| Enum default | Validates enums serialize as numbers by default |
| Enum -EnumsAsStrings | Validates -EnumsAsStrings parameter |
| null | Validates null serialization |
| Arrays | Validates array serialization |
| Hashtable | Validates ordered hashtable serialization |
| Nested objects | Validates nested PSCustomObject serialization |
| Default escaping | Validates no escaping by default |
| -EscapeHandling EscapeHtml | Validates HTML character escaping |
| -EscapeHandling EscapeNonAscii | Validates non-ASCII escaping |
| -Depth | Validates Depth parameter |
| -AsArray | Validates AsArray parameter |
| Multiple pipeline objects | Validates array output for multiple inputs |
| Depth over 100 | Validates depth limit enforcement |

### CODE REVIEW

**Comment by iSazonov** [comment]
I guess we can add more tests while working on V2 JSON serializer. So I will keep this open some time.

**Comment by iSazonov** [comment]
If there are duplicates we could simply add comments in existing tests to describe additional scenarios are tested. If we can not create clear comments please keep duplicate tests.

**Comment by yotsuda** [comment]
@iSazonov Sorry for the initial submission with redundant tests. I've refactored based on your feedback:

**Removed** (covered by existing tests):
- null, arrays, hashtable (lines 94-107)
- Depth, AsArray, Multiple pipeline (lines 43-77)

**Consolidated** into TestCases format:
- Enum: default + `-EnumsAsStrings`
- EscapeHandling: EscapeHtml + EscapeNonAscii

**Kept**:
- Uri, Nested PSCustomObject (renamed), EscapeHandling default (renamed), Depth over 100 error

**Added**:
- Special floating-point values (`Infinity`, `-Infinity`, `NaN`)
- Empty array (`$null`)
- SwitchParameter (`{"IsPresent":true/false}`)

Result: +29/-51 lines, removing redundancy while improving edge case coverage.

I'll add more test cases if I find other scenarios worth covering while working on V2 improvements.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1 [hunk 1]
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
...
+        $json = @{ text = "<>&" } | ConvertTo-Json -Compress
+        $json | Should -BeExactly '{"text":"<>&"}'
+    }
+
+    It 'Should escape HTML with -EscapeHandling EscapeHtml' {
+        $json = @{ text = "<>&" } | ConvertTo-Json -Compress -EscapeHandling EscapeHtml
+        $json | Should -Match '\\u003C'
+        $json | Should -Match '\\u003E'
+        $json | Should -Match '\\u0026'
+    }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: Kept, but consolidated into TestCases format. [Line 87](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L87) tests apostrophe escaping (`'`), while this tests HTML tag characters (`<>&`). Different character sets require explicit coverage.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1 [hunk 2]
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
...
+    It 'Depth parameter should work' {
+        $obj = @{ a = @{ b = 1 } }
+        $json = $obj | ConvertTo-Json -Depth 2 -Compress
+        $json | Should -BeExactly '{"a":{"b":1}}'
+    }
+
+    It 'AsArray parameter should work' {
+        $json = @{a=1} | ConvertTo-Json -AsArray -Compress
+        $json | Should -BeExactly '[{"a":1}]'
+    }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: Removed. Covered by [lines 71-77](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L71-L77).

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1 [hunk 3]
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
...
+
+    It 'Multiple objects from pipeline should be serialized as array' {
+        $json = 1, 2, 3 | ConvertTo-Json -Compress
+        $json | Should -BeExactly '[1,2,3]'
+    }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: Removed. Covered by [lines 75-76](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L75-L76).

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1 [hunk 4]
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
...
+    It 'Should escape non-ASCII with -EscapeHandling EscapeNonAscii' {
+        $json = @{ text = "日本語" } | ConvertTo-Json -Compress -EscapeHandling EscapeNonAscii
+        $json | Should -Match '\\u'
+    }
+
+    It 'Depth parameter should work' {
+        $obj = @{ a = @{ b = 1 } }
+        $json = $obj | ConvertTo-Json -Depth 2 -Compress
+        $json | Should -BeExactly '{"a":{"b":1}}'
+    }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: Removed. Covered by [lines 43-68](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L43-L68).

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1 [hunk 5]
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
...
+
+    It 'Should serialize enums as strings with -EnumsAsStrings' {
+        $json = [System.DayOfWeek]::Monday | ConvertTo-Json -EnumsAsStrings
+        $json | Should -BeExactly '"Monday"'
+    }
+
+    It 'Should serialize null correctly' {
+        $null | ConvertTo-Json | Should -BeExactly 'null'
+    }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: Removed. Covered by [line 96](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L96).

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1 [hunk 6]
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
...
+            }
+        }
+        $json = $obj | ConvertTo-Json -Compress
+        $json | Should -BeExactly '{"name":"test","child":{"value":42}}'
+    }
+
+    It 'Should not escape by default' {
+        $json = @{ text = "<>&" } | ConvertTo-Json -Compress
+        $json | Should -BeExactly '{"text":"<>&"}'
+    }
```
[the remark this answers is not in the document]
Comment (yotsuda) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: Kept with renamed test. [Line 85](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L85) tests apostrophe default behavior, while this tests HTML tag characters (`<>&`) default behavior. Since EscapeHtml specifically targets HTML characters, explicitly testing their default (non-escaped) behavior adds value.

**Comment by iSazonov** [comment]
@yotsuda  Have we tests to cover all code paths for `if (currentDepth > context.MaxDepth)`? Notice, we don't add extended/adapted properties if isPurePSObj = true; and add otherwise.

**Comment by yotsuda** [comment]
@iSazonov Added tests covering both code paths for `currentDepth > context.MaxDepth` (b10f8be):

1. **Pure PSObject** (`isPurePSObj = true`): `AddPsProperties` returns early without adding ETS properties. The `ToString()` result is used as-is.
2. **Non-pure PSObject** (`isPurePSObj = false`): `AddPsProperties` processes ETS properties on the string-converted value.

```powershell
# Pure PSObject - depth exceeded (isPurePSObj = true)
$pso = [PSCustomObject]@{ Name = 'test' }
$pso | Add-Member -NotePropertyName ETSProp -NotePropertyValue 'ets'
@{ inner = $pso } | ConvertTo-Json -Depth 0 -Compress
# Output: {"inner":"@{Name=test; ETSProp=ets}"}

# Non-pure PSObject (Version) - depth exceeded (isPurePSObj = false)
$version = [version]'1.2.3'
$version | Add-Member -NotePropertyName ETSProp -NotePropertyValue 'ets'
@{ inner = $version } | ConvertTo-Json -Depth 0 -Compress
# Output: {"inner":"1.2.3"}
```

Also added a comparison test for pure PSObject serialization within depth limit.

Also fixed the DateTime ETS test (f3025f7) to use UTC, avoiding timezone conversion issues that caused intermittent failures.

**Comment by iSazonov** [comment]
@yotsuda Thanks! I see if object is pure PSObject LanguagePrimitives.ConvertTo() does all work and add ETS properties so we skipp AddPSProperty(). But I wonder why these properties are not added if isPurePSObj == false - I'd expect AddPSProperty() should add. Thoughts?

**Comment by yotsuda** [comment]
@iSazonov Let me trace through the code flow (referring to V1/JsonObject.cs).

When `isPurePSObj == false` and `currentDepth > MaxDepth`:

1. The object is converted to string via `LanguagePrimitives.ConvertTo(obj, typeof(string), ...)` ([line 618-619](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L618-L619))
2. `AddPsProperties` is called with the converted string
3. Since `isPurePSObj == false`, we don't early-return at [line 671-673](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L671-L673)
4. The string is wrapped in a new Dictionary with key "value" ([line 680-682](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L680-L682))
5. `AppendPsProperties` is called ([line 685](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L685))

In `AppendPsProperties`:
- [Line 710-712](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L710-L712) checks if the **original** `psObj.BaseObject` is string or DateTime - if so, it returns early (no ETS properties added)
- [Line 716-719](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L716-L719): Since `isCustomObj == false`, only Extended properties are searched (not Adapted)
- Any Extended properties found are added to the dictionary

After `AppendPsProperties` returns:
- [Line 687-690](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L687-L690): If the original `rv` was not a dictionary and no properties were added (dict.Count == 1), the original string is returned instead of the dictionary wrapper

So the ETS properties **are** added for `isPurePSObj == false` when:
- The original BaseObject is not string/DateTime, AND
- Extended properties exist on the PSObject

The properties are **not** added when:
- BaseObject is string/DateTime (intentional, these types use their natural serialization)
- No Extended properties exist

---

**Note for #26654 (V2/STJ):** While testing this, I found an issue in `WriteProperty` where `MaxDepth == 0` directly called `ToString()` without going through `WriteDepthExceeded`, so ETS properties on nested PSObjects were not included. I've fixed this in #26654 (commit 67a1bbd40) and added a test to cover this case.

**Comment by iSazonov** [comment]
@yotsuda Please look failed tests.

I think it is time to switch to the PR and add more compatibility tests. I will share my thoughts as you will ready.

**Comment by yotsuda** [comment]
@iSazonov Fixed the cross-platform test failures by copying the test file from PR #26654 (feature-convertto-json-factory-v2).
The test file is now identical to the one in #26654.

**Comment by iSazonov** [comment]
@yotsuda It would be a pity to lose the experience that we gained while exploring the possibility of migrating to STJ.
If you have a desire to improve the ConvertTo-Json cmdlet, then you could start with tests.
I looked at what we have now. In fact, very little, poor test coverage.
Therefore, we could use the capabilities of AI and generate a complete set of tests so as to eliminate any regression with subsequent changes.
Below are my thoughts. Maybe I missed something.

Object types:
1. Primitive/scalar types + specific cases (string, datetime - ETS; BigInt; IPAddress?; URI?; NaN for double and so on)
2. Enum (+ enums as strings)
2. Arrays
3. Dictionary
4. PSCustomObject
5. Composite object (I'd prefer PowerShell class instead of FileInfo)

The following options are needed for each type of object:
1. Pipeline vs InputObject (e.g. PSObject wrapped vs unwrapped)
2. With ETS vs no ETS properties

Additionally, tests are needed for scenarios (most likely it is more convenient to combine them):
1. Depth truncation
2. Multilevel compositions

By Multilevel compositions, I mean that, for example, the elements of an array can be scalars, enums, array, dictionary, PSCustomObject and Composite object; the same applies to all other composite types.

Of course, for all combinations, this would be a huge set of tests. Therefore, they should be reduced reasonably. So we need to check all scalars, but for composite types it makes no sense to check all scalars - you can limit the list to string, datetime, int, enum plus nested composite types with the same restrictions.
Even in this case, it's likely that there will be too many tests for one PR.
It may be worth starting with scalars, as well as scalars as elements of other types.
Then array and dictionary. Then PSCustomObject and Composite object. Then Depth truncation and Multilevel compositions.

**Comment by yotsuda** [comment]
@iSazonov Thank you for sharing this comprehensive test plan! I completely agree that having thorough test coverage is essential before any changes to `ConvertTo-Json`.

## Implementation

I've implemented Phase 1 (scalars + scalars as elements) in commit 735c6cf66. The tests cover:

| Category | Tests | Coverage |
|----------|-------|----------|
| Primitive types | 14 | int, long, double, float, decimal, bool |
| String types | 9 | regular, empty, special chars, unicode, emoji |
| DateTime types | 5 | UTC, Local, Unspecified, DateTimeOffset, TimeSpan |
| Guid | 3 | Pipeline vs InputObject behavior |
| Uri | 4 | http, https, query, file schemes |
| Enum | 11 | numeric, EnumsAsStrings, flags |
| IPAddress | 3 | IPv4, IPv6, Extended properties |
| Pipeline vs InputObject | 7 | consistency verification |
| Scalars in arrays | 9 | including mixed types |
| Scalars in hashtables | 8 | hashtable and PSCustomObject |
| ETS on scalars | 4 | string/DateTime ignored, Uri/Guid included |

**Total: 77 new tests**

I plan to continue adding the remaining phases (Array, Dictionary, PSCustomObject, Composite, Depth truncation) in this same PR, unless you'd prefer separate PRs.

## Interesting findings

While writing tests, I discovered some behaviors worth documenting:
- `Guid` via Pipeline includes Extended property, via InputObject serializes as string only
- `TimeSpan` serializes as object with all properties (not ISO string)
- `double` 0.0 serializes as `"0.0"` not `"0"`

---

All my PRs to this repo have been fully automated via [PowerShell.MCP](https://github.com/yotsuda/PowerShell.MCP) - investigating existing behavior, coding, building, and testing without writing a single line myself. Claude Code couldn't handle PowerShell this well.

The bigger picture: MCP connects AI to external tools, and countless purpose-specific MCP servers are being developed - [Azure MCP Server](https://github.com/microsoft/mcp), [GitHub MCP Server](https://github.com/github/github-mcp-server), etc. PowerShell.MCP makes all of them obsolete - Az module is far more capable than Azure MCP Server, git.exe/gh.exe far more than GitHub MCP Server. AI can simply use these mature, full-featured tools directly through PowerShell. This positions PowerShell as the universal AI interface - I hope this could be good news for the PowerShell community. Feedback welcome!

cc: @mklement0

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1 [hunk 7]
```diff
@@ -156,4 +173,633 @@ Describe 'ConvertTo-Json' -tags "CI" {
...
+            $result = $outer | ConvertTo-Json -Depth 0 -WarningAction SilentlyContinue
+            $parsed = $result | ConvertFrom-Json
+
+            $parsed.File.CustomExt | Should -BeExactly 'extended_value'
+            $parsed.File.value | Should -BeLike '*System.Management.Automation.dll'
+        }
+
+    }
+
+    #region Comprehensive Scalar Type Tests (Phase 1)
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: There are too many new tests above. We can get confused. I think it will be easier for us if we make a separate PR for each phase.

Comment (yotsuda) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1]: Agreed. I think the comprehensive tests should be in separate PRs to make review easier.

I've restructured this PR:

1. **This PR (#26639)** - Existing tests only:
   - Core behavior tests (null handling, ETS properties, EscapeHandling)
   - Depth exceeded behavior tests (isPurePSObj distinction)
   - Nested raw object serialization tests

2. **Separate PR (to follow)** - Phase 1 scalar type tests:
   - Primitive types (int, long, double, bool, etc.)
   - String types (escape sequences, unicode, emoji)
   - DateTime and related types
   - Guid, Uri, Enum, IPAddress
   - Pipeline vs InputObject behavior
   - Scalars in arrays and hashtables
   - ETS properties on scalar types

I'll follow up with separate PRs for subsequent phases.
